### PR TITLE
Suggest block pattern transformations that are contextual to the currently selected 'simple' blocks (no InnerBlocks)

### PR DESCRIPTION
**Pull request by ntsekouras**
## Description

Resolves partially: #28736
Part of: #27575

This PR is a spinoff from: https://github.com/WordPress/gutenberg/pull/29890 for suggesting block pattern transformations that are contextual to the currently selected blocks.

The goal of this PR is to handle 'simple' blocks (without InnerBlocks) first, to better manage the complexity and arbitrariness. This is a complex problem where we need to find the right balance between magic(opinionated actions) and APIs. 

## Notes
1. The core blocks that have been updated to include `role:content` in some of their attributes are:
`core/paragraph`, `core/heading`, `core/list`, `core/quote`, `core/pullquote`, `core/preformatted`, `core/verse`, `core/search`.

2. For patterns to use this transformations need to have declared `blockTypes` property, so it now uses only my test patterns. See: https://github.com/WordPress/gutenberg/pull/30471.

## Rationale behind the current decisions
1. I have skipped any check about transforming selected blocks to available block transforms inside patterns, as I think it would be very confusing to have a really different suggested pattern. An example would be to have a paragraph selected and show patterns that do not include paragraph but have transformed the paragraph to a quote block.
2. To show a suggested pattern, **every selected block must have matched with a pattern's block** . This doesn't necessarily mean that the order will be preserved. You can test this in this PR by selecting a paragraph and two headings (in this order) and see the suggested pattern that has the following order: [heading, paragraph, heading]. Also shown at the gif.
3. If a suggested pattern has more blocks than the selected blocks, add them as are from the pattern. Every pattern block that matched with a selected block will be transformed. IMO this is the main reason for doing this and having block patterns in general. We want to inspire/show users alternative designs.

## APIs
1. The block pattern filtering uses the experimental `blockTypes` API introduced here: https://github.com/WordPress/gutenberg/pull/30471
2. A new `__experimentalRole` property is added to block attribute's API. This property is optional and can mark attributes based on their role. Here is utilized by having `content` value and an example [is the `Heading` block](https://github.com/WordPress/gutenberg/pull/30469/files#diff-801846106160a8e52c94f9316006260c2e6438ba5e6ee086b93d9f5f074122bcR14). This is used by for knowing which attributes we need to retain during the pattern transformation, so as the user can keep their block's content and make use of the styling/presentation attributes of the block pattern.

I have many comments in my code trying to explain as good as I can the flows.

## Testing instructions
1. Test with the demo patterns I created in lib/test-block-patterns.php or register one yourselves
2. Create some paragraphs or headings and make different selections (single or multiple) to observe the previews from the selected blocks.
3. Only if all of the selected blocks can be matched in pattern's blocks, it will be shown as an option

## Screenshots 
simple transforms

## TODO in follow ups
- [ ] More tests for some functions + e2e.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

File: lib/test-block-patterns.php
Comment (youknowriad) [inline lib/test-block-patterns.php]: Is this really needed? why not just show all the patterns in the inserter?

File: lib/test-block-patterns.php
Comment (youknowriad) [inline lib/test-block-patterns.php]: Is this like a shortcut to avoid parsing the pattern? Should it be just `blockTypes` at the top level of the pattern?

File: lib/test-block-patterns.php
Comment (ntsekouras) [reply lib/test-block-patterns.php]: In these patterns it's not needed and all these will be removed before merge. `inserter` is used in `Query` ([check here](https://github.com/WordPress/gutenberg/pull/28891/files#diff-f811d58d1bc263104d65422eb480a2a629a86775e98e7522ecdb95f69ad606c6R17)) which have `content` to be placed in InnerBlocks. This behaviour might change though with all explorations of patterns and integration in different flows (placeholder, transforms etc..).

File: lib/test-block-patterns.php
Comment (ntsekouras) [reply lib/test-block-patterns.php]: I've created this PR to remove `scope`: https://github.com/WordPress/gutenberg/pull/30471

**Comment by jasmussen** [comment]
This is what I see:

status

This has enormous potential, truly, and it works very seamlessly. There are a great deal of followups we need to make:

- The "Patterns" submenu and the whole transformation dialog could use a revamp knowing what's now possible.
- The fact that it's only Paragraphs and Headings that have patterns to transform into, I suspect that's limited in part by availability of patterns, correct? Or are there other limits imposed?
- Many patterns are wrapped in a group. If I transform a pattern inside a group into a pattern that adds yet another group, should we be smarter about removing one of those group wrappers? My instinct is no.
- But related to the group thing: should I be able to transform a group? Or would that simply let me transform into anything arbitrary because groups are used so widely?

These are very ux-heavy questions, and as a first iteration, I'm happy to approve what's here, because the implementation is does not weigh down what we have. But this also assumes that we _are_ able to make the subsequent changes. For example, @jameskoster has some mockups around transforming root level blocks using a carousel — I presume that could use the same underlying tech here?

**Comment by jameskoster** [comment]
> @jameskoster has some mockups around transforming root level blocks using a carousel

For context, I assume you're referring to this one:

https://user-images.githubusercontent.com/846565/114021467-a4923c80-9868-11eb-9d5e-f6959a0f4a3c.mp4

Disclaimer: This is still a very rough concept :D 

---

> But related to the group thing: should I be able to transform a group?

This is an interesting question. Since Groups have very little (no?) semantic meaning in this context, I can see a world where the patterns exposed while transforming a Group are identical to the ones that would be exposed by transforming a selection of the children of that Group.

**Comment by ntsekouras** [comment]
Thanks for reviewing/testing!

>The fact that it's only Paragraphs and Headings that have patterns to transform into, I suspect that's limited in part by availability of patterns, correct?

For patterns to use this `transformations` need to have declared `blockTypes`, so it now uses only my test patterns.
Also I need to add the `role:content` attribute to other blocks as well in this PR.

>Many patterns are wrapped in a group. If I transform a pattern inside a group into a pattern that adds yet another group, should we be smarter about removing one of those group wrappers? My instinct is no.

For start I think so too, that we shouldn't try to be too smart, although we have this small check in `Group` block. I think it can be explored in follow up PRs that will need to handle blocks with `InnerBlocks`, unless we find some issues with the testing now :).

> But related to the group thing: should I be able to transform a group? Or would that simply let me transform into anything arbitrary because groups are used so widely?

For now, blocks with `InnerBlocks` are not handled and when they will be, I think `Group` and `Columns` at least, should not be handled at all, as they are `grouping` blocks where the user could have anything inside as content.


>I'm happy to approve what's here, because the implementation is does not weigh down what we have. But this also assumes that we are able to make the subsequent changes.

The logic is here and we are getting `transformed previews`, so we could possibly just change the UI that shows them in the future.

**Comment by jasmussen** [comment]
> For patterns to use this transformations need to have declared blockTypes, so it now uses only my test patterns.
> Also I need to add the role:content attribute to other blocks as well in this PR.

Gotcha. So it seems like to land this we need to graduate to a set of specific patterns though. Seems okay to start small, but would be good to start well. If it's Headings and Paragraphs right now, perhaps we could ask @kjellr for advice on what patterns might be good to surface as potential transformations?

**Comment by ntsekouras** [comment]
>Seems okay to start small, but would be good to start well.

Agreed and since this is part of all pattern integration explorations we should add just a few..

> If it's Headings and Paragraphs right now...

It will be more blocks (ex list). I'm checking now and push with a couple of new test patterns.

**Comment by ntsekouras** [comment]
I have updated the other core blocks that can utilize `role:content` in some of their attributes. See the PRs description for details.

I have also included a demo pattern for `search` block which seems to me a nice use case for this feature. @kjellr

**Comment by jameskoster** [comment]
Fixed a small padding issue I noticed. Otherwise I agree with Joen, this feels like the first step towards something much bigger and more exciting, and due to the small footprint in the UI it is good to merge and iterate on.

**Comment by kjellr** [comment]
This looks cool! 

For the feature itself: I'll just note that from a casual user perspective, seeing this really blurs the lines between a "block style" and a "block pattern". The difference is purely a technical one, and I don't think users will know that at all. 

In terms of the patterns though, I think we should keep it as minimal as possible in this initial PR — maybe just use a paragraph or header example or two? I should have a little time tomorrow to help narrow in on that if needed. @melchoyce and I are going to start thinking through a more cohesive set of single-block patterns for this feature separately. I've opened https://github.com/WordPress/gutenberg/issues/30631 to start tracking that work.

**Comment by kjellr** [comment]
> In terms of the patterns though, I think we should keep it as minimal as possible in this initial PR — maybe just use a paragraph or header example or two? 

After thinking about it overnight, I'm not entirely sure we need to bundle any patterns as part of this PR yet. There are a lot of new patterns coming in https://github.com/WordPress/gutenberg/pull/29973 that will work really well here. And as noted above, we'll brainstorm more of these in https://github.com/WordPress/gutenberg/issues/30631. 

But if folks do think they're useful, here are just a couple "single block" patterns that I think are ok if we want to do that. The first is a simple customized paragraph, and the others are featuring relatively new settings for their respective blocks — I think patterns could be a way to help introduce those to folks. 

**Large Paragraph with Background Color**

Screen Shot 2021-04-09 at 11 16 19 AM

```
<!-- wp:paragraph {"style":{"color":{"link":"#FFFFFF","text":"#FFFFFF","background":"#000000"},"typography":{"lineHeight":"1.3","fontSize":"26px"}}} -->
<p class="has-text-color has-background has-link-color" style="--wp--style--color--link:#FFFFFF;background-color:#000000;color:#FFFFFF;font-size:26px;line-height:1.3">The whole series of my life appeared to me as a dream; I sometimes doubted if indeed it were all true, for it never presented itself to my mind with the force of reality.</p>
<!-- /wp:paragraph -->
```

**Social links with a shared background color**

Screen Shot 2021-04-09 at 11 34 00 AM


```
<!-- wp:social-links {"customIconColor”:”#ffffff”,”iconColorValue”:”#ffffff,”customIconBackgroundColor":"#3962e3","iconBackgroundColorValue":"#3962e3"} -->
<ul class="wp-block-social-links has-icon-color has-icon-background-color"><!-- wp:social-link {"url":"https://wordpress.org","service":"wordpress"} /-->
<!-- wp:social-link {"url":"#","service":"chain"} /-->
<!-- wp:social-link {"url":"#","service":"mail"} /--></ul>
<!-- /wp:social-links -->
```

**Two full-width buttons**

Screen Shot 2021-04-09 at 11 46 23 AM

```
<!-- wp:buttons -->
<div class="wp-block-buttons"><!-- wp:button {"borderRadius":4,"style":{"color":{"text":"#ffffff","background":"#3962e3"}},"width":100} -->
<div class="wp-block-button has-custom-width wp-block-button__width-100"><a class="wp-block-button__link has-text-color has-background" style="border-radius:4px;background-color:#3962e3;color:#ffffff">Learn More</a></div>
<!-- /wp:button -->
<!-- wp:button {"borderRadius":4,"style":{"color":{"background":"#3962e3","text":"#ffffff"}},"width":100} -->
<div class="wp-block-button has-custom-width wp-block-button__width-100"><a class="wp-block-button__link has-text-color has-background" style="border-radius:4px;background-color:#3962e3;color:#ffffff">Get Started</a></div>
<!-- /wp:button --></div>
<!-- /wp:buttons -->
```

File: packages/blocks/src/api/test/utils.js
Comment (mcsf) [inline packages/blocks/src/api/test/utils.js]: My obsession is fixating on the fact that this `res` variable stands out from the other test cases. :P

File: packages/block-editor/src/components/block-switcher/index.js
Comment (mcsf) [inline packages/block-editor/src/components/block-switcher/index.js]: This can now be passed directly in the return object, no need for the temp var: `{ patterns: __experimentalGet... `

File: packages/block-editor/src/store/selectors.js
Comment (mcsf) [inline packages/block-editor/src/store/selectors.js]:
```suggestion
		 * For now we only handle blocks without InnerBlocks and take into account
		 * the `role` property of blocks' attributes for the transformation.  Note
		 * that the blocks have been retrieved through `getBlock`, which doesn't
		 * return the inner blocks of an inner block controller, so we still need
		 * to check for this case too.
		 */
```

File: packages/block-editor/src/store/selectors.js
Comment (mcsf) [inline packages/block-editor/src/store/selectors.js]: Needs update to reflect simpler approach.

File: packages/block-editor/src/store/selectors.js
Comment (mcsf) [inline packages/block-editor/src/store/selectors.js:1895]: Perhaps a nitpick, but this selector returns _just_ patterns. That is, at the top of the call stack for this selector we find `state.settings.__experimentalBlockPatterns`. What I mean is that "pattern transform items" doesn't refer to a type different from "pattern", correct?

I suppose that the name came out of symmetry with the preexisting logic in `BlockSwitcher`, in particular with `getBlockTransformItems`. This isn't something for now, but I wouldn't mind revisiting the nomenclature for all these assorted selectors. Just as the prefix `get` has a meaning (returns an object, whereas `has` returns a boolean), the main noun in the name should reflect the return type. So we might consider something like `getPatternsForBlocks`, `getMatchingBlockPatterns`, `getAvailableBlockPatterns`, etc.

File: packages/block-library/src/preformatted/block.json
Comment (mcsf) [inline packages/block-library/src/preformatted/block.json]: Any thoughts on whether to go with `role` or `__experimentalRole`?

File: lib/load.php
Comment (mcsf) [inline lib/load.php]: Don't forget to remove these before merging!

File: packages/block-editor/src/components/block-switcher/use-transformed-patterns.js
Comment (mcsf) [inline packages/block-editor/src/components/block-switcher/use-transformed-patterns.js]: Technically this is not quite `WPBlockPattern[]`, right? We're adding `transformedBlocks` as a property.

**Review by mcsf — APPROVED** [review]
Looks good, thanks for all the work here!

I left a bunch of open questions, but they shouldn't block this PR.

File: packages/block-library/src/preformatted/block.json
Comment (ntsekouras) [reply packages/block-library/src/preformatted/block.json]: We can go with `__experimentalRole`, yes.

File: packages/block-editor/src/store/selectors.js
Comment (ntsekouras) [reply packages/block-editor/src/store/selectors.js:1895]: I see your point. We can revisit on follow ups for this functionality.

**Review by kjellr — APPROVED** [review]
This works well in my testing! I tried out the included patterns, and also a handful of the ones in https://github.com/WordPress/gutenberg/pull/29973 (with a temporary `blockTypes` prop added), and they all worked as expected. 👍

**Comment by mcsf** [comment]
Great job here, @ntsekouras!

**Pull request merged**
